src/arch/arm/kernel/boot.c

```diff
@@ -303,8 +304,12 @@ BOOT_CODE static void release_secondary_cpus(void)
 
     /* Wait until all the secondary cores are done initialising */
     while (ksNumCPUs != CONFIG_MAX_NUM_NODES) {
-        /* perform a memory release+acquire to get new values of ksNumCPUs */
-        __atomic_signal_fence(__ATOMIC_ACQ_REL);
+        /* perform a memory acquire to get new values of ksNumCPUs, release for ksCurTime */
+        __atomic_thread_fence(__ATOMIC_ACQ_REL);
+#if defined(CONFIG_DEBUG_BUILD) && defined(CONFIG_KERNEL_MCS)
+        /* Test whether clocks are synchronised across nodes: */
+        NODE_STATE(ksCurTime) = getCurrentTime();
```

Comment:
It's nitpicking, but there must be a `__atomic_thread_fence(__ATOMIC_RELEASE)` afterwards if we want to ensure the value becomes visible exactly here in the last loop iteration also. Otherwise, the last iteration has no pipeline barrier, as it just executed the check in the loop header and then leaves the function

Reply:
That's why `clock_sync_test()` is called before incrementing `ksNumCPUs`, to guarantee that `ksCurTime` on node 0 will keep changing.

Reply:
I'd still recommend using this order, ie insert the new code before the barrier.

```
while (ksNumCPUs != CONFIG_MAX_NUM_NODES) {
#if defined(CONFIG_DEBUG_BUILD) && defined(CONFIG_KERNEL_MCS)
        /* Test whether clocks are synchronised across nodes: */
        NODE_STATE(ksCurTime) = getCurrentTime();
#endif
        /* ensure new values of ksNumCPUs are read and ksCurTime is written */
        __atomic_thread_fence(__ATOMIC_ACQ_REL);
}